ReSharper.FSharp/src/FSharp.Psi/src/Util/FSharpExpressionUtil.cs

```diff
@@ -35,5 +35,106 @@ public static class FSharpExpressionUtil
 
     public static readonly Func<IFSharpExpression, bool> IsSimpleValueExpressionFunc = IsSimpleValueExpression;
     public static readonly Func<IFSharpExpression, bool> IsLiteralExpressionFunc = IsLiteralExpression;
+
+    // TODO: change name
+    public static IFSharpExpression GetOuterMostParentExpression(this IFSharpExpression expression)
+    {
+      IFSharpTreeNode currentExpr = expression;
+      var parent = currentExpr.Parent;
+      while (parent != null)
+      {
+        switch (parent)
+        {
+          case IChameleonExpression:
+            return currentExpr as IFSharpExpression;
+          case ILiteralExpr literalExpr:
+            currentExpr = literalExpr;
+            parent = literalExpr.Parent;
+            break;
+          case IMatchClause matchClause:
+            if (matchClause.Expression == currentExpr)
+            {
+              currentExpr = matchClause;
+              parent = matchClause.Parent;
+            }
```

Comment:
This looks a bit like reimplementation of `MatchClauseNavigator.GetByExpression` and similar methods. For going up the tree they are preferred way in many cases, since they allow skipping intermediate nodes (when the rule allows) and make it easier to change the tree structure and grammar.

With multi-step navigators used, so it could be done in a way like this:
```csharp
while (true)
{
  if (MatchExprNavigator.GetByClauseExpression(currentExpr) is { } matchExpr)
  {
    currentExpr = matchExpr;
    continue;
  }

  if (SequentialExprNavigator.GetByExpression(currentExpr) is { } seqExpr &&
      seqExpr.ExpressionsEnumerable.Last() == currentExpr)
  {
    currentExpr = seqExpr;
    continue;
  }

  // other navigators go 

  return currentExpr;
}
```

This way you don't have to deal with non-expr parent nodes (like the chameleon expr case above). This is also arguably clearer than a big switch statement. 🙂

Getting parent multiple times is not a problem performance-wise here, as it's a simple field lookup and we only run this code when a caret is at a matching error highlighting type.